# Titanic competition with TensorFlow Decision Forests

This notebook will take you through the steps needed to train a baseline Gradient Boosted Trees Model using TensorFlow Decision Forests and creating a submission on the Titanic competition. 

This notebook shows:

1. How to do some basic pre-processing. For example, the passenger names will be tokenized, and ticket names will be splitted in parts.
1. How to train a Gradient Boosted Trees (GBT) with default parameters
1. How to train a GBT with improved default parameters
1. How to tune the parameters of a GBTs
1. How to train and ensemble many GBTs

# Imports dependencies

In [1]:
import numpy as np
import pandas as pd
import os

import tensorflow as tf
import tensorflow_decision_forests as tfdf

print(f"Found TF-DF {tfdf.__version__}")

Found TF-DF 1.2.0


# Load dataset

In [2]:
train_df = pd.read_csv("/kaggle/input/titanic/train.csv")
serving_df = pd.read_csv("/kaggle/input/titanic/test.csv")

train_df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

# Prepare dataset

We will apply the following transformations on the dataset.

1. Tokenize the names. For example, "Braund, Mr. Owen Harris" will become ["Braund", "Mr.", "Owen", "Harris"].
2. Extract any prefix in the ticket. For example ticket "STON/O2. 3101282" will become "STON/O2." and 3101282.

In [3]:
def preprocess(df):
    df = df.copy()
    
    def normalize_name(x):
        return " ".join([v.strip(",()[].\"'") for v in x.split(" ")])
    
    def ticket_number(x):
        return x.split(" ")[-1]
        
    def ticket_item(x):
        items = x.split(" ")
        if len(items) == 1:
            return "NONE"
        return "_".join(items[0:-1])
    
    df["Name"] = df["Name"].apply(normalize_name)
    df["Ticket_number"] = df["Ticket"].apply(ticket_number)
    df["Ticket_item"] = df["Ticket"].apply(ticket_item)                     
    return df
    
preprocessed_train_df = preprocess(train_df)
preprocessed_serving_df = preprocess(serving_df)

preprocessed_train_df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                              Name     Sex   Age  SibSp  \
0                            Braund Mr Owen Harris    male  22.0      1   
1  Cumings Mrs John Bradley Florence Briggs Thayer  female  38.0      1   
2                             Heikkinen Miss Laina  female  26.0      0   
3         Futrelle Mrs Jacques Heath Lily May Peel  female  35.0      1   
4                           Allen Mr William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked Ticket_number Ticket_item  
0      0         A/5 21171   7.2500   NaN        S         21171         A/5  
1      0          PC 17599  71.2833   C85        C         17599          PC  
2      0  STON/O2. 3101282   7.9250   NaN        S       3101282    STON/O2.  
3      0            113803  53.1000  C123        S        113803        NONE  
4      0            373450   8.0500   NaN        S        373450        NONE

Let's keep the list of the input features of the model. Notably, we don't want to train our model on the "PassengerId" and "Ticket" features.

In [4]:
input_features = list(preprocessed_train_df.columns)
input_features.remove("Ticket")
input_features.remove("PassengerId")
input_features.remove("Survived")
#input_features.remove("Ticket_number")

print(f"Input features: {input_features}")

Input features: ['Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin', 'Embarked', 'Ticket_number', 'Ticket_item']


# Convert Pandas dataset to TensorFlow Dataset

In [5]:
def tokenize_names(features, labels=None):
    """Divite the names into tokens. TF-DF can consume text tokens natively."""
    features["Name"] =  tf.strings.split(features["Name"])
    return features, labels

train_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_train_df,label="Survived").map(tokenize_names)
serving_ds = tfdf.keras.pd_dataframe_to_tf_dataset(preprocessed_serving_df).map(tokenize_names)

# Train model with default parameters

### Train model

First, we are training a GradientBoostedTreesModel model with the default parameters.

In [6]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    random_seed=1234,
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-09-09T02:42:03.305052562+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmdg24njg/model/ with prefix 73ae476925524afd
[INFO 2024-09-09T02:42:03.315494153+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:42:03.315627072+00:00 kernel.cc:1046] Use fast generic engine


Please report this to the TensorFlow team. When filing the bug, set the verbosity to 10 (on Linux, `export AUTOGRAPH_VERBOSITY=10`) and attach the full output.
Cause: could not get source code
To silence this warning, decorate the function with @tf.autograph.experimental.do_not_convert
Accuracy: 0.8260869383811951 Loss:0.8608942627906799


# Train model with improved default parameters

Now you'll use some specific parameters when creating the GBT model

In [7]:
model = tfdf.keras.GradientBoostedTreesModel(
    verbose=0, # Very few logs
    features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
    exclude_non_specified_features=True, # Only use the features in "features"
    
    #num_trees=2000,
    
    # Only for GBT.
    # A bit slower, but great to understand the model.
    # compute_permutation_variable_importance=True,
    
    # Change the default hyper-parameters
    # hyperparameter_template="benchmark_rank1@v1",
    
    #num_trees=1000,
    #tuner=tuner
    
    min_examples=1,
    categorical_algorithm="RANDOM",
    #max_depth=4,
    shrinkage=0.05,
    #num_candidate_attributes_ratio=0.2,
    split_axis="SPARSE_OBLIQUE",
    sparse_oblique_normalization="MIN_MAX",
    sparse_oblique_num_projections_exponent=2.0,
    num_trees=2000,
    #validation_ratio=0.0,
    random_seed=1234,
    
)
model.fit(train_ds)

self_evaluation = model.make_inspector().evaluation()
print(f"Accuracy: {self_evaluation.accuracy} Loss:{self_evaluation.loss}")

[INFO 2024-09-09T02:42:08.1828186+00:00 kernel.cc:1214] Loading model from path /tmp/tmp8hzfzodj/model/ with prefix 91c7eb78f9254ee9
[INFO 2024-09-09T02:42:08.193220018+00:00 decision_forest.cc:661] Model loaded with 33 root(s), 1823 node(s), and 10 input feature(s).
[INFO 2024-09-09T02:42:08.193299139+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.760869562625885 Loss:1.0154211521148682


Let's look at the model and you can also notice the information about variable importance that the model figured out

In [8]:
model.summary()

Model: "gradient_boosted_trees_model_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
Total params: 1
Trainable params: 0
Non-trainable params: 1
_________________________________________________________________
Type: "GRADIENT_BOOSTED_TREES"
Task: CLASSIFICATION
Label: "__LABEL"

Input Features (11):
	Age
	Cabin
	Embarked
	Fare
	Name
	Parch
	Pclass
	Sex
	SibSp
	Ticket_item
	Ticket_number

No weights

Variable Importance: INV_MEAN_MIN_DEPTH:
    1.           "Sex"  0.576632 ################
    2.           "Age"  0.364297 #######
    3.          "Fare"  0.278839 ####
    4.          "Name"  0.208548 #
    5. "Ticket_number"  0.180792 
    6.        "Pclass"  0.176962 
    7.         "Parch"  0.176659 
    8.   "Ticket_item"  0.175540 
    9.      "Embarked"  0.172339 
   10.         "SibSp"  0.170442 

Variable Importance: NUM_AS_ROOT:
    1.  "Sex" 28.000000 ################
    2. "Name"  5.000000 

# Make predictions

In [9]:
def prediction_to_kaggle_format(model, threshold=0.5):
    proba_survive = model.predict(serving_ds, verbose=0)[:,0]
    return pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (proba_survive >= threshold).astype(int)
    })

def make_submission(kaggle_predictions):
    path="/kaggle/working/submission.csv"
    kaggle_predictions.to_csv(path, index=False)
    print(f"Submission exported to {path}")
    
kaggle_predictions = prediction_to_kaggle_format(model)
make_submission(kaggle_predictions)
!head /kaggle/working/submission.csv

Submission exported to /kaggle/working/submission.csv
PassengerId,Survived
892,0
893,0
894,0
895,0
896,0
897,0
898,0
899,0
900,1


# Training a model with hyperparameter tunning

Hyper-parameter tuning is enabled by specifying the tuner constructor argument of the model. The tuner object contains all the configuration of the tuner (search space, optimizer, trial and objective).


In [10]:
tuner = tfdf.tuner.RandomSearch(num_trials=1000)
tuner.choice("min_examples", [2, 5, 7, 10])
tuner.choice("categorical_algorithm", ["CART", "RANDOM"])

local_search_space = tuner.choice("growing_strategy", ["LOCAL"])
local_search_space.choice("max_depth", [3, 4, 5, 6, 8])

global_search_space = tuner.choice("growing_strategy", ["BEST_FIRST_GLOBAL"], merge=True)
global_search_space.choice("max_num_nodes", [16, 32, 64, 128, 256])

#tuner.choice("use_hessian_gain", [True, False])
tuner.choice("shrinkage", [0.02, 0.05, 0.10, 0.15])
tuner.choice("num_candidate_attributes_ratio", [0.2, 0.5, 0.9, 1.0])


tuner.choice("split_axis", ["AXIS_ALIGNED"])
oblique_space = tuner.choice("split_axis", ["SPARSE_OBLIQUE"], merge=True)
oblique_space.choice("sparse_oblique_normalization",
                     ["NONE", "STANDARD_DEVIATION", "MIN_MAX"])
oblique_space.choice("sparse_oblique_weights", ["BINARY", "CONTINUOUS"])
oblique_space.choice("sparse_oblique_num_projections_exponent", [1.0, 1.5])

# Tune the model. Notice the `tuner=tuner`.
tuned_model = tfdf.keras.GradientBoostedTreesModel(tuner=tuner)
tuned_model.fit(train_ds, verbose=0)

tuned_self_evaluation = tuned_model.make_inspector().evaluation()
print(f"Accuracy: {tuned_self_evaluation.accuracy} Loss:{tuned_self_evaluation.loss}")

Use /tmp/tmpqs40r14y as temporary training directory


[INFO 2024-09-09T02:44:17.699163037+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqs40r14y/model/ with prefix 5f4c83bbd4854b7d
[INFO 2024-09-09T02:44:17.715847137+00:00 decision_forest.cc:661] Model loaded with 19 root(s), 589 node(s), and 12 input feature(s).
[INFO 2024-09-09T02:44:17.715931726+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesGeneric" built
[INFO 2024-09-09T02:44:17.715983693+00:00 kernel.cc:1046] Use fast generic engine


Accuracy: 0.9178082346916199 Loss:0.6503586769104004


In the last line in the cell above, you can see the accuracy is higher than previously with default parameters and parameters set by hand.

This is the main idea behing hyperparameter tuning.

For more information you can follow this tutorial: [Automated hyper-parameter tuning](https://www.tensorflow.org/decision_forests/tutorials/automatic_tuning_colab)

# Making an ensemble

Here you'll create 100 models with different seeds and combine their results

This approach removes a little bit the random aspects related to creating ML models

In the GBT creation is used the `honest` parameter. It will use different training examples to infer the structure and the leaf values. This regularization technique trades examples for bias estimates.

In [11]:
predictions = None
num_predictions = 0

for i in range(100):
    print(f"i:{i}")
    # Possible models: GradientBoostedTreesModel or RandomForestModel
    model = tfdf.keras.GradientBoostedTreesModel(
        verbose=0, # Very few logs
        features=[tfdf.keras.FeatureUsage(name=n) for n in input_features],
        exclude_non_specified_features=True, # Only use the features in "features"

        #min_examples=1,
        #categorical_algorithm="RANDOM",
        ##max_depth=4,
        #shrinkage=0.05,
        ##num_candidate_attributes_ratio=0.2,
        #split_axis="SPARSE_OBLIQUE",
        #sparse_oblique_normalization="MIN_MAX",
        #sparse_oblique_num_projections_exponent=2.0,
        #num_trees=2000,
        ##validation_ratio=0.0,
        random_seed=i,
        honest=True,
    )
    model.fit(train_ds)
    
    sub_predictions = model.predict(serving_ds, verbose=0)[:,0]
    if predictions is None:
        predictions = sub_predictions
    else:
        predictions += sub_predictions
    num_predictions += 1

predictions/=num_predictions

kaggle_predictions = pd.DataFrame({
        "PassengerId": serving_df["PassengerId"],
        "Survived": (predictions >= 0.5).astype(int)
    })

make_submission(kaggle_predictions)

i:0


[INFO 2024-09-09T02:44:18.87557925+00:00 kernel.cc:1214] Loading model from path /tmp/tmpczh7sbbe/model/ with prefix a1b617ee22b54f79
[INFO 2024-09-09T02:44:18.880622308+00:00 kernel.cc:1046] Use fast generic engine


i:1


[INFO 2024-09-09T02:44:20.526973058+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4tgcxgpj/model/ with prefix 577f0d1e9fdb407f
[INFO 2024-09-09T02:44:20.54816069+00:00 kernel.cc:1046] Use fast generic engine


i:2


[INFO 2024-09-09T02:44:21.699134666+00:00 kernel.cc:1214] Loading model from path /tmp/tmp55equ3ng/model/ with prefix 86f795734cd84ef4
[INFO 2024-09-09T02:44:21.704488976+00:00 kernel.cc:1046] Use fast generic engine


i:3


[INFO 2024-09-09T02:44:23.497624271+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsr67vfao/model/ with prefix 6f2edb5f263e4a1f
[INFO 2024-09-09T02:44:23.527542717+00:00 kernel.cc:1046] Use fast generic engine


i:4


[INFO 2024-09-09T02:44:24.677560853+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3ni64fba/model/ with prefix 9a2df103dcfd41ee
[INFO 2024-09-09T02:44:24.6838945+00:00 kernel.cc:1046] Use fast generic engine


i:5


[INFO 2024-09-09T02:44:25.776950197+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqwubd8a8/model/ with prefix 9d6a28d6825c465f
[INFO 2024-09-09T02:44:25.780614176+00:00 kernel.cc:1046] Use fast generic engine


i:6


[INFO 2024-09-09T02:44:26.987791639+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnkn926gz/model/ with prefix c381e7e6da584300
[INFO 2024-09-09T02:44:26.996183543+00:00 kernel.cc:1046] Use fast generic engine


i:7


[INFO 2024-09-09T02:44:28.607317908+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdhfzrql5/model/ with prefix b078fb689aa8420d
[INFO 2024-09-09T02:44:28.629515311+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:44:28.629566884+00:00 kernel.cc:1046] Use fast generic engine


i:8


[INFO 2024-09-09T02:44:29.900775619+00:00 kernel.cc:1214] Loading model from path /tmp/tmpk4pnlqyy/model/ with prefix a654305a6d0244ec
[INFO 2024-09-09T02:44:29.911879496+00:00 kernel.cc:1046] Use fast generic engine


i:9


[INFO 2024-09-09T02:44:31.818357116+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvzj8j3mm/model/ with prefix 445a699c6f08443b
[INFO 2024-09-09T02:44:31.836348294+00:00 kernel.cc:1046] Use fast generic engine


i:10


[INFO 2024-09-09T02:44:33.02279077+00:00 kernel.cc:1214] Loading model from path /tmp/tmpuhfie75v/model/ with prefix 64fcefc54a2646eb
[INFO 2024-09-09T02:44:33.02946748+00:00 kernel.cc:1046] Use fast generic engine


i:11


[INFO 2024-09-09T02:44:34.497317751+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu1seb7yh/model/ with prefix 77bb8e9922804802
[INFO 2024-09-09T02:44:34.51412855+00:00 kernel.cc:1046] Use fast generic engine


i:12


[INFO 2024-09-09T02:44:35.691542275+00:00 kernel.cc:1214] Loading model from path /tmp/tmpccx_u49z/model/ with prefix 056937807a0b4189
[INFO 2024-09-09T02:44:35.698395049+00:00 kernel.cc:1046] Use fast generic engine


i:13


[INFO 2024-09-09T02:44:37.060318653+00:00 kernel.cc:1214] Loading model from path /tmp/tmpgjeknfb7/model/ with prefix a718db9bd92e43d3
[INFO 2024-09-09T02:44:37.072765787+00:00 kernel.cc:1046] Use fast generic engine


i:14


[INFO 2024-09-09T02:44:38.256172401+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu2vya8dj/model/ with prefix 23953afd465644a3
[INFO 2024-09-09T02:44:38.263214685+00:00 kernel.cc:1046] Use fast generic engine


i:15


[INFO 2024-09-09T02:44:39.465869878+00:00 kernel.cc:1214] Loading model from path /tmp/tmpha8_nfoh/model/ with prefix 1ec1373811c74b98
[INFO 2024-09-09T02:44:39.47475184+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:44:39.474803384+00:00 kernel.cc:1046] Use fast generic engine


i:16


[INFO 2024-09-09T02:44:40.832302196+00:00 kernel.cc:1214] Loading model from path /tmp/tmparokmkba/model/ with prefix e0bcb69dbb554f82
[INFO 2024-09-09T02:44:40.847177385+00:00 kernel.cc:1046] Use fast generic engine


i:17


[INFO 2024-09-09T02:44:42.244125873+00:00 kernel.cc:1214] Loading model from path /tmp/tmpe39dvghj/model/ with prefix c602438eb1aa4e9e
[INFO 2024-09-09T02:44:42.258968233+00:00 kernel.cc:1046] Use fast generic engine


i:18


[INFO 2024-09-09T02:44:43.597410351+00:00 kernel.cc:1214] Loading model from path /tmp/tmpp5l6gt8n/model/ with prefix 3b9c855da6094d51
[INFO 2024-09-09T02:44:43.610690907+00:00 kernel.cc:1046] Use fast generic engine


i:19


[INFO 2024-09-09T02:44:45.139636529+00:00 kernel.cc:1214] Loading model from path /tmp/tmpizqh4_xf/model/ with prefix ac2746c4a2584d76
[INFO 2024-09-09T02:44:45.15964011+00:00 kernel.cc:1046] Use fast generic engine


i:20


[INFO 2024-09-09T02:44:46.691508264+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4zljepnd/model/ with prefix 0bd8b8db39d24c2b
[INFO 2024-09-09T02:44:46.708832355+00:00 kernel.cc:1046] Use fast generic engine


i:21


[INFO 2024-09-09T02:44:47.884472759+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqs2i9mig/model/ with prefix b5874097b9b941d6
[INFO 2024-09-09T02:44:47.890553714+00:00 kernel.cc:1046] Use fast generic engine


i:22


[INFO 2024-09-09T02:44:49.063365088+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9lv902ls/model/ with prefix eee9e6bfd7174d1f
[INFO 2024-09-09T02:44:49.070132336+00:00 kernel.cc:1046] Use fast generic engine


i:23


[INFO 2024-09-09T02:44:50.378913157+00:00 kernel.cc:1214] Loading model from path /tmp/tmpu16o0k61/model/ with prefix 3fd4156a27e9403c
[INFO 2024-09-09T02:44:50.388942235+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:44:50.389007086+00:00 kernel.cc:1046] Use fast generic engine


i:24


[INFO 2024-09-09T02:44:51.57704207+00:00 kernel.cc:1214] Loading model from path /tmp/tmpdaohm49y/model/ with prefix 09a9065b501246e2
[INFO 2024-09-09T02:44:51.583658674+00:00 kernel.cc:1046] Use fast generic engine


i:25


[INFO 2024-09-09T02:44:52.948577383+00:00 kernel.cc:1214] Loading model from path /tmp/tmp37qufbtp/model/ with prefix e59ea5fd817347be
[INFO 2024-09-09T02:44:52.961734676+00:00 kernel.cc:1046] Use fast generic engine


i:26


[INFO 2024-09-09T02:44:54.275240103+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2y414gr9/model/ with prefix ec49f40809514604
[INFO 2024-09-09T02:44:54.287448102+00:00 kernel.cc:1046] Use fast generic engine


i:27


[INFO 2024-09-09T02:44:55.475370717+00:00 kernel.cc:1214] Loading model from path /tmp/tmpedn6vqsa/model/ with prefix debb3ccfa0c74673
[INFO 2024-09-09T02:44:55.482906422+00:00 kernel.cc:1046] Use fast generic engine


i:28


[INFO 2024-09-09T02:44:56.694747881+00:00 kernel.cc:1214] Loading model from path /tmp/tmpybh0btvp/model/ with prefix 2963bf77e579494a
[INFO 2024-09-09T02:44:56.700661471+00:00 kernel.cc:1046] Use fast generic engine


i:29


[INFO 2024-09-09T02:44:58.155235624+00:00 kernel.cc:1214] Loading model from path /tmp/tmpr1qvljmj/model/ with prefix 21399db97fd64620
[INFO 2024-09-09T02:44:58.170229516+00:00 kernel.cc:1046] Use fast generic engine


i:30


[INFO 2024-09-09T02:45:00.06469113+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvnqv1hc3/model/ with prefix 73174461f6864543
[INFO 2024-09-09T02:45:00.094652573+00:00 kernel.cc:1046] Use fast generic engine


i:31


[INFO 2024-09-09T02:45:02.102350794+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9zhgb5pz/model/ with prefix fa18a468ab064727
[INFO 2024-09-09T02:45:02.115748754+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:45:02.115799386+00:00 kernel.cc:1046] Use fast generic engine


i:32


[INFO 2024-09-09T02:45:03.397765365+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnp7mo0xo/model/ with prefix dcfef70575c54b32
[INFO 2024-09-09T02:45:03.405018097+00:00 kernel.cc:1046] Use fast generic engine


i:33


[INFO 2024-09-09T02:45:04.891781646+00:00 kernel.cc:1214] Loading model from path /tmp/tmptjuuzlm6/model/ with prefix 8002caa63ee9445e
[INFO 2024-09-09T02:45:04.908829304+00:00 kernel.cc:1046] Use fast generic engine


i:34


[INFO 2024-09-09T02:45:06.260113124+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzib__1vv/model/ with prefix bd0a16eec555476c
[INFO 2024-09-09T02:45:06.271166035+00:00 kernel.cc:1046] Use fast generic engine


i:35


[INFO 2024-09-09T02:45:07.547264504+00:00 kernel.cc:1214] Loading model from path /tmp/tmpowv7f3hb/model/ with prefix 7c32f5ea82db4d4c
[INFO 2024-09-09T02:45:07.55799211+00:00 kernel.cc:1046] Use fast generic engine


i:36


[INFO 2024-09-09T02:45:09.017482524+00:00 kernel.cc:1214] Loading model from path /tmp/tmphtvuu66e/model/ with prefix fbf085bcf8fc47b5
[INFO 2024-09-09T02:45:09.033820407+00:00 kernel.cc:1046] Use fast generic engine


i:37


[INFO 2024-09-09T02:45:10.339692035+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6va4jmle/model/ with prefix d041339db7cc49f7
[INFO 2024-09-09T02:45:10.349704633+00:00 kernel.cc:1046] Use fast generic engine


i:38


[INFO 2024-09-09T02:45:11.806237759+00:00 kernel.cc:1214] Loading model from path /tmp/tmp90xny0ha/model/ with prefix df105b09fe6849f5
[INFO 2024-09-09T02:45:11.821226291+00:00 kernel.cc:1046] Use fast generic engine


i:39


[INFO 2024-09-09T02:45:13.236670811+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0bv7_pvs/model/ with prefix 1a1310d3c0774c0e
[INFO 2024-09-09T02:45:13.251278861+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:45:13.251348164+00:00 kernel.cc:1046] Use fast generic engine


i:40


[INFO 2024-09-09T02:45:14.355323339+00:00 kernel.cc:1214] Loading model from path /tmp/tmp6e_mfo1j/model/ with prefix 9d4ecd52a20b4f3c
[INFO 2024-09-09T02:45:14.360405642+00:00 kernel.cc:1046] Use fast generic engine


i:41


[INFO 2024-09-09T02:45:15.848659541+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2wqc44wp/model/ with prefix 433c5520de9445dc
[INFO 2024-09-09T02:45:15.866142397+00:00 kernel.cc:1046] Use fast generic engine


i:42


[INFO 2024-09-09T02:45:17.157847404+00:00 kernel.cc:1214] Loading model from path /tmp/tmpw4j6dok7/model/ with prefix 169c07332dae4fd9
[INFO 2024-09-09T02:45:17.168190029+00:00 kernel.cc:1046] Use fast generic engine


i:43


[INFO 2024-09-09T02:45:18.705337758+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvcmaz526/model/ with prefix 1f27c5ff652b487d
[INFO 2024-09-09T02:45:18.724731777+00:00 kernel.cc:1046] Use fast generic engine


i:44


[INFO 2024-09-09T02:45:20.002202027+00:00 kernel.cc:1214] Loading model from path /tmp/tmptpca8x9u/model/ with prefix 3b2b21958c7f4f1c
[INFO 2024-09-09T02:45:20.013293421+00:00 kernel.cc:1046] Use fast generic engine


i:45


[INFO 2024-09-09T02:45:21.120120087+00:00 kernel.cc:1214] Loading model from path /tmp/tmpsmgnceye/model/ with prefix 48a7cfe3a6be44e6
[INFO 2024-09-09T02:45:21.124338303+00:00 kernel.cc:1046] Use fast generic engine


i:46


[INFO 2024-09-09T02:45:22.687626793+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvnnh52y1/model/ with prefix fa568f7da3db4adb
[INFO 2024-09-09T02:45:22.704372839+00:00 kernel.cc:1046] Use fast generic engine


i:47


[INFO 2024-09-09T02:45:24.156395193+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1cmnnth9/model/ with prefix 1022b22745a5461a
[INFO 2024-09-09T02:45:24.171398019+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:45:24.171468627+00:00 kernel.cc:1046] Use fast generic engine


i:48


[INFO 2024-09-09T02:45:25.298714719+00:00 kernel.cc:1214] Loading model from path /tmp/tmp906znwc4/model/ with prefix fbc4fb9e2bd14e99
[INFO 2024-09-09T02:45:25.303781769+00:00 kernel.cc:1046] Use fast generic engine


i:49


[INFO 2024-09-09T02:45:26.538713654+00:00 kernel.cc:1214] Loading model from path /tmp/tmplz7napld/model/ with prefix 1662a641ab5b4268
[INFO 2024-09-09T02:45:26.546444592+00:00 kernel.cc:1046] Use fast generic engine


i:50


[INFO 2024-09-09T02:45:27.909412363+00:00 kernel.cc:1214] Loading model from path /tmp/tmptllgpq01/model/ with prefix 3f307a4323fa4f54
[INFO 2024-09-09T02:45:27.923165407+00:00 kernel.cc:1046] Use fast generic engine


i:51


[INFO 2024-09-09T02:45:29.42604575+00:00 kernel.cc:1214] Loading model from path /tmp/tmpyl04je7j/model/ with prefix ea351c9443734a92
[INFO 2024-09-09T02:45:29.443523798+00:00 kernel.cc:1046] Use fast generic engine


i:52


[INFO 2024-09-09T02:45:30.690016051+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbrsmms2d/model/ with prefix 5259860a701a4412
[INFO 2024-09-09T02:45:30.698959765+00:00 kernel.cc:1046] Use fast generic engine


i:53


[INFO 2024-09-09T02:45:31.94744153+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2c1lvuqh/model/ with prefix 6a31ebaef16b406f
[INFO 2024-09-09T02:45:31.955914099+00:00 kernel.cc:1046] Use fast generic engine


i:54


[INFO 2024-09-09T02:45:33.096095471+00:00 kernel.cc:1214] Loading model from path /tmp/tmptk9sbiyn/model/ with prefix 7397e204d8e04343
[INFO 2024-09-09T02:45:33.100008461+00:00 kernel.cc:1046] Use fast generic engine


i:55


[INFO 2024-09-09T02:45:34.560219559+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvlqs0t18/model/ with prefix 6c509a315b584d26
[INFO 2024-09-09T02:45:34.577685508+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:45:34.577740405+00:00 kernel.cc:1046] Use fast generic engine


i:56


[INFO 2024-09-09T02:45:36.689137337+00:00 kernel.cc:1214] Loading model from path /tmp/tmp13q83v0q/model/ with prefix 1a1decb7fa9b43b5
[INFO 2024-09-09T02:45:36.704331312+00:00 kernel.cc:1046] Use fast generic engine


i:57


[INFO 2024-09-09T02:45:37.960096975+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzm4s3bg5/model/ with prefix 2b2c8116ffef407d
[INFO 2024-09-09T02:45:37.965305509+00:00 kernel.cc:1046] Use fast generic engine


i:58


[INFO 2024-09-09T02:45:39.211865258+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3z28280o/model/ with prefix 510c3f1fdc334f6e
[INFO 2024-09-09T02:45:39.219556806+00:00 kernel.cc:1046] Use fast generic engine


i:59


[INFO 2024-09-09T02:45:40.549112621+00:00 kernel.cc:1214] Loading model from path /tmp/tmpmtsct7kd/model/ with prefix 016f18e282754a17
[INFO 2024-09-09T02:45:40.559887915+00:00 kernel.cc:1046] Use fast generic engine


i:60


[INFO 2024-09-09T02:45:41.887564495+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzeubl145/model/ with prefix 1e4550cb786d41e9
[INFO 2024-09-09T02:45:41.899625231+00:00 kernel.cc:1046] Use fast generic engine


i:61


[INFO 2024-09-09T02:45:43.079560693+00:00 kernel.cc:1214] Loading model from path /tmp/tmpez8m50qp/model/ with prefix 12e9407144c54990
[INFO 2024-09-09T02:45:43.085416086+00:00 kernel.cc:1046] Use fast generic engine


i:62


[INFO 2024-09-09T02:45:44.948792384+00:00 kernel.cc:1214] Loading model from path /tmp/tmpm3czjob8/model/ with prefix 514f73abfb634911
[INFO 2024-09-09T02:45:44.977986193+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:45:44.978041006+00:00 kernel.cc:1046] Use fast generic engine


i:63


[INFO 2024-09-09T02:45:46.342702189+00:00 kernel.cc:1214] Loading model from path /tmp/tmp1qod195s/model/ with prefix e65cfd0ab7f94036
[INFO 2024-09-09T02:45:46.353128248+00:00 kernel.cc:1046] Use fast generic engine


i:64


[INFO 2024-09-09T02:45:47.641879341+00:00 kernel.cc:1214] Loading model from path /tmp/tmpnecwmuf2/model/ with prefix 9e192cdb31b040d7
[INFO 2024-09-09T02:45:47.650643523+00:00 kernel.cc:1046] Use fast generic engine


i:65


[INFO 2024-09-09T02:45:48.818915055+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxvr2cvv6/model/ with prefix 6bc54c9158a44695
[INFO 2024-09-09T02:45:48.82470047+00:00 kernel.cc:1046] Use fast generic engine


i:66


[INFO 2024-09-09T02:45:50.052554209+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3noozaf2/model/ with prefix 4bd3dec265e14b3b
[INFO 2024-09-09T02:45:50.05996849+00:00 kernel.cc:1046] Use fast generic engine


i:67


[INFO 2024-09-09T02:45:51.659119169+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkhzzqrff/model/ with prefix 0242902f52a14f5e
[INFO 2024-09-09T02:45:51.680893296+00:00 kernel.cc:1046] Use fast generic engine


i:68


[INFO 2024-09-09T02:45:53.093787861+00:00 kernel.cc:1214] Loading model from path /tmp/tmptf5p5oqe/model/ with prefix a5d70e913bf14f97
[INFO 2024-09-09T02:45:53.107138039+00:00 kernel.cc:1046] Use fast generic engine


i:69


[INFO 2024-09-09T02:45:54.394264275+00:00 kernel.cc:1214] Loading model from path /tmp/tmp2q94br0i/model/ with prefix 17e99d7470e34972
[INFO 2024-09-09T02:45:54.401943495+00:00 kernel.cc:1046] Use fast generic engine


i:70


[INFO 2024-09-09T02:45:55.726322761+00:00 kernel.cc:1214] Loading model from path /tmp/tmppo3elcgb/model/ with prefix c42c6289859641e1
[INFO 2024-09-09T02:45:55.739694639+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:45:55.739750768+00:00 kernel.cc:1046] Use fast generic engine


i:71


[INFO 2024-09-09T02:45:57.031868845+00:00 kernel.cc:1214] Loading model from path /tmp/tmpf7hvk_z9/model/ with prefix 5b7a91f8fb08417d
[INFO 2024-09-09T02:45:57.040522096+00:00 kernel.cc:1046] Use fast generic engine


i:72


[INFO 2024-09-09T02:45:58.558036524+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4v7dayvz/model/ with prefix 54599d96d12b4e72
[INFO 2024-09-09T02:45:58.576012765+00:00 kernel.cc:1046] Use fast generic engine


i:73


[INFO 2024-09-09T02:45:59.850516457+00:00 kernel.cc:1214] Loading model from path /tmp/tmp5p4bjbix/model/ with prefix cf033a5d5bcb4191
[INFO 2024-09-09T02:45:59.858912464+00:00 kernel.cc:1046] Use fast generic engine


i:74


[INFO 2024-09-09T02:46:01.246169493+00:00 kernel.cc:1214] Loading model from path /tmp/tmpkpa1hhkb/model/ with prefix 832392bfe4054918
[INFO 2024-09-09T02:46:01.259005845+00:00 kernel.cc:1046] Use fast generic engine


i:75


[INFO 2024-09-09T02:46:02.53450407+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9keefltb/model/ with prefix cd43a824b0a24377
[INFO 2024-09-09T02:46:02.544065276+00:00 kernel.cc:1046] Use fast generic engine


i:76


[INFO 2024-09-09T02:46:03.689717448+00:00 kernel.cc:1214] Loading model from path /tmp/tmp4kj46cnc/model/ with prefix 44e4eefba63a4e9e
[INFO 2024-09-09T02:46:03.694213226+00:00 kernel.cc:1046] Use fast generic engine


i:77


[INFO 2024-09-09T02:46:04.847865655+00:00 kernel.cc:1214] Loading model from path /tmp/tmpzjqbnb6i/model/ with prefix e4b6606ed70a4d32
[INFO 2024-09-09T02:46:04.853220193+00:00 kernel.cc:1046] Use fast generic engine


i:78


[INFO 2024-09-09T02:46:06.079956548+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwyk_hd4_/model/ with prefix fca99218fdf64c89
[INFO 2024-09-09T02:46:06.087588069+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:46:06.087647821+00:00 kernel.cc:1046] Use fast generic engine


i:79


[INFO 2024-09-09T02:46:07.306058438+00:00 kernel.cc:1214] Loading model from path /tmp/tmp9ebiy5ot/model/ with prefix d5a5735d78bd4753
[INFO 2024-09-09T02:46:07.31385002+00:00 kernel.cc:1046] Use fast generic engine


i:80


[INFO 2024-09-09T02:46:08.619938663+00:00 kernel.cc:1214] Loading model from path /tmp/tmpcmo30f58/model/ with prefix e313bd721d344942
[INFO 2024-09-09T02:46:08.630835182+00:00 kernel.cc:1046] Use fast generic engine


i:81


[INFO 2024-09-09T02:46:09.999237287+00:00 kernel.cc:1214] Loading model from path /tmp/tmpiukol3hi/model/ with prefix a49c4da60e7d406d
[INFO 2024-09-09T02:46:10.012092316+00:00 kernel.cc:1046] Use fast generic engine


i:82


[INFO 2024-09-09T02:46:11.334442825+00:00 kernel.cc:1214] Loading model from path /tmp/tmpop42140t/model/ with prefix 0991263d425c4724
[INFO 2024-09-09T02:46:11.345797616+00:00 kernel.cc:1046] Use fast generic engine


i:83


[INFO 2024-09-09T02:46:12.615153679+00:00 kernel.cc:1214] Loading model from path /tmp/tmp3g3_hac3/model/ with prefix da230fe5bc144f68
[INFO 2024-09-09T02:46:12.6249153+00:00 kernel.cc:1046] Use fast generic engine


i:84


[INFO 2024-09-09T02:46:14.960659674+00:00 kernel.cc:1214] Loading model from path /tmp/tmpqnxotb0j/model/ with prefix deb2d56964db4ec7
[INFO 2024-09-09T02:46:14.982516344+00:00 kernel.cc:1046] Use fast generic engine


i:85


[INFO 2024-09-09T02:46:16.34318294+00:00 kernel.cc:1214] Loading model from path /tmp/tmp7nrfjxar/model/ with prefix e1f56428ef5e4dca
[INFO 2024-09-09T02:46:16.351750053+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:46:16.351801525+00:00 kernel.cc:1046] Use fast generic engine


i:86


[INFO 2024-09-09T02:46:17.932524782+00:00 kernel.cc:1214] Loading model from path /tmp/tmpleuoe3_o/model/ with prefix df081aa54ab04fed
[INFO 2024-09-09T02:46:17.951758099+00:00 kernel.cc:1046] Use fast generic engine


i:87


[INFO 2024-09-09T02:46:19.548112315+00:00 kernel.cc:1214] Loading model from path /tmp/tmpvxl4p9d1/model/ with prefix 4ed4e734e75443d8
[INFO 2024-09-09T02:46:19.568484449+00:00 kernel.cc:1046] Use fast generic engine


i:88


[INFO 2024-09-09T02:46:20.970824067+00:00 kernel.cc:1214] Loading model from path /tmp/tmpl5khjnwa/model/ with prefix 0857df9cf43c49e1
[INFO 2024-09-09T02:46:20.984299817+00:00 kernel.cc:1046] Use fast generic engine


i:89


[INFO 2024-09-09T02:46:22.173143457+00:00 kernel.cc:1214] Loading model from path /tmp/tmpaefy6wkp/model/ with prefix 3a974b23255046e8
[INFO 2024-09-09T02:46:22.178881231+00:00 kernel.cc:1046] Use fast generic engine


i:90


[INFO 2024-09-09T02:46:23.546445853+00:00 kernel.cc:1214] Loading model from path /tmp/tmp48e68l0z/model/ with prefix 8c1d7793ed974a23
[INFO 2024-09-09T02:46:23.557932675+00:00 kernel.cc:1046] Use fast generic engine


i:91


[INFO 2024-09-09T02:46:24.778723083+00:00 kernel.cc:1214] Loading model from path /tmp/tmpbw_5v355/model/ with prefix 40dcb7fffb0b45a5
[INFO 2024-09-09T02:46:24.786527969+00:00 kernel.cc:1046] Use fast generic engine


i:92


[INFO 2024-09-09T02:46:26.437091949+00:00 kernel.cc:1214] Loading model from path /tmp/tmp51wbre6_/model/ with prefix dacc227b1d9b4e52
[INFO 2024-09-09T02:46:26.457453142+00:00 abstract_model.cc:1311] Engine "GradientBoostedTreesQuickScorerExtended" built
[INFO 2024-09-09T02:46:26.457524048+00:00 kernel.cc:1046] Use fast generic engine


i:93


[INFO 2024-09-09T02:46:27.809770956+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoqb5lh9o/model/ with prefix 501dfee5f7334eb3
[INFO 2024-09-09T02:46:27.821119749+00:00 kernel.cc:1046] Use fast generic engine


i:94


[INFO 2024-09-09T02:46:29.037337267+00:00 kernel.cc:1214] Loading model from path /tmp/tmp0l74v6ht/model/ with prefix 06a0df16465d4d6f
[INFO 2024-09-09T02:46:29.044628509+00:00 kernel.cc:1046] Use fast generic engine


i:95


[INFO 2024-09-09T02:46:30.336112032+00:00 kernel.cc:1214] Loading model from path /tmp/tmpoynbwrs3/model/ with prefix 22bbbbfa4d2e4219
[INFO 2024-09-09T02:46:30.345508672+00:00 kernel.cc:1046] Use fast generic engine


i:96


[INFO 2024-09-09T02:46:31.690137962+00:00 kernel.cc:1214] Loading model from path /tmp/tmpb72pzz0v/model/ with prefix 47693436cbc3471d
[INFO 2024-09-09T02:46:31.701714473+00:00 kernel.cc:1046] Use fast generic engine


i:97


[INFO 2024-09-09T02:46:32.875334049+00:00 kernel.cc:1214] Loading model from path /tmp/tmph0pqdlmi/model/ with prefix 26eada62f3c94930
[INFO 2024-09-09T02:46:32.881791268+00:00 kernel.cc:1046] Use fast generic engine


i:98


[INFO 2024-09-09T02:46:34.165886084+00:00 kernel.cc:1214] Loading model from path /tmp/tmpxp8_c7q_/model/ with prefix 40b2018b8885433a
[INFO 2024-09-09T02:46:34.17494948+00:00 kernel.cc:1046] Use fast generic engine


i:99


[INFO 2024-09-09T02:46:35.667874854+00:00 kernel.cc:1214] Loading model from path /tmp/tmpwauw6wv4/model/ with prefix e90b2282a56944fe
[INFO 2024-09-09T02:46:35.684213866+00:00 kernel.cc:1046] Use fast generic engine


Submission exported to /kaggle/working/submission.csv


# What is next

If you want to learn more about TensorFlow Decision Forests and its advanced features, you can follow the official documentation [here](https://www.tensorflow.org/decision_forests) 